orchestrator/models.py

```diff
@@ -58,17 +90,48 @@ class Job(BaseModel):
     walltime = models.IntegerField(blank=True, null=True, default=None)
     memlimit = models.CharField(blank=True, null=True, default=None, max_length=20)
 
-    def save(self, *args, **kwargs):
-        if self.status != Status.CREATED:
-            if not self.submitted:
-                self.submitted = self.created_date
-            if self.status != Status.PENDING:
-                if not self.started:
-                    self.started = self.created_date
-        if self.status == Status.COMPLETED or self.status == Status.FAILED:
-            if not self.finished:
-                self.finished = self.modified_date
-        super().save(*args, **kwargs)
+    def submit_to_lsf(self, external_id, job_store_dir, job_work_dir, job_output_dir, log_path):
+        self.status = Status.SUBMITTED
+        self.external_id = external_id
+        self.job_store_location = job_store_dir
+        self.working_dir = job_work_dir
+        self.output_directory = job_output_dir
+        self.log_path = log_path
+        self.submitted = timezone.now()
+        self.message['log'] = log_path
+        self.save(update_fields=['status',
+                                 'external_id',
+                                 'job_store_location',
+                                 'working_dir',
+                                 'output_directory',
+                                 'submitted',
+                                 'message'])
+
+    def update_status(self, lsf_status):
+        if self.status == Status.PENDING and lsf_status == Status.RUNNING:
+            self.started = timezone.now()
+        self.status = lsf_status
+        self.save(update_fields=['status', 'started'])
```

Comment:
I don't see where `update_status` is called, but I'm worried about calling `Status.RUNNING` multiple times and setting `self.started = now` each time

Reply:
To that point, I'm not sure the FSM should allow for a state to transition to itself in the case of `RUNNING`/`PENDING`

Reply:
I assume PENDING can transition to RUNNING only once and that would be the moment when execution is really started. First the Job is SUBMITTED, and then by `CHECK_STATUS_ON_LSF` we pick up the state from LSF and it becomes PENDING. Since we don't know when the LSF scheduler will be able to run, it can stay in PENDING for few cycles. After LSF have the resources and move it to RUNNING we will set the `self.started` and transition to RUNNING. Job is expected to stay in RUNNING for some time, but can't go back to PENDING anymore. Is that correct or I am missing something?

Reply:
No that's good! In that case, I'm not 100% sure about `SUBMITTING`, but I get the impression it'll make a lot more sense when I see the code.